packages/@aws-cdk/cfn-include/lib/cdk-include.ts

```diff
@@ -0,0 +1,221 @@
+ * You can retrieve any of the resources defined in the template file by calling the {@link getResource} method.
+ * Any modifications you make on the returned resource objects will be reflected in the resulting CDK template.
+ *
+ * @see getResource
+ */
+export class CdkInclude extends core.Construct {
+  private readonly resources: { [logicalId: string]: core.CfnResource } = {};
+  private readonly cfnType2L1Class: { [cfnType: string]: string };
+  private readonly template: any;
+  private readonly preserveLogicalIds: boolean;
+
+  constructor(scope: core.Construct, id: string, props: CdkIncludeProps) {
+    super(scope, id);
+
+    // read the template into a JS object
+    this.template = fs.readJsonSync(props.templateFile);
+
+    // check for any sections of the template we don't recognize
+    for (const section of Object.keys(this.template)) {
+      switch (section) {
+        case 'AWSTemplateFormatVersion':
+        case 'Description':
+        case 'Resources':
+          // currently, we only support these 3 sections
+          break;
+        default:
+          throw new Error(`Encountered an unsupported section in the template: '${section}'`);
+      }
+    }
+
+    // read the map of CloudFormation type to their L1 class created by the build script
+    this.cfnType2L1Class = fs.readJsonSync(path.join(__dirname, '..', 'cfn-types-2-classes.json'));
+    this.preserveLogicalIds = props.preserveLogicalIds ?? true;
+
+    // instantiate all resources as CDK L1 objects
+    for (const logicalId of Object.keys(this.template.Resources || {})) {
+      this.getOrCreateResource(logicalId);
+    }
+  }
+
+  /**
+   * Returns the low-level CfnResource from the template with the given logical ID.
+   * Any modifications performed on that resource will be reflected in the resulting CDK template.
+   *
+   * The returned object will be of the proper underlying class;
+   * you can always cast it to the correct type in your code:
+   *
+   *     // assume the template contains an AWS::S3::Bucket with logical ID 'Bucket'
+   *     const cfnBucket = cfnTemplate.getResource('Bucket') as s3.CfnBucket;
+   *     // cfnBucket is of type s3.CfnBucket
+   *
+   * If the template does not contain a resource with the given logical ID,
+   * an exception will be thrown.
+   *
+   * @param logicalId the logical ID of the resource in the CloudFormation template file
+   */
+  public getResource(logicalId: string): core.CfnResource {
+    const ret = this.resources[logicalId];
+    if (!ret) {
+      throw new Error(`Resource with logical ID '${logicalId}' was not found in the template`);
+    }
+    return ret;
+  }
+
+  private getOrCreateResource(logicalId: string): core.CfnResource {
+    const ret = this.resources[logicalId];
+    if (ret) {
+      return ret;
+    }
+
+    const resourceConfig: any = this.template.Resources[logicalId];
+    const l1ClassFqn = this.cfnType2L1Class[(resourceConfig as any).Type];
```

Comment:
I am wondering why we need this mapping at all? There is a simple deterministic function that takes a CloudFormation resource type and produces the CfnClass name from it. Are you trying to future proof against changes in this logic?

Reply:
2 reasons:

1. It's not only about the class name, it's also about the CDK package. For example, I'm not sure how to deduce that `AWS::Serverless::Api` is mapped to be `@aws-cdk/aws-sam.CfnApi`.
2. This allows us to force a dependency on modules that contain L1s. Without it, it's actually very difficult to know which packages in `@aws-cdk/*` this module should depend on.

Reply:
Kk got it